### PR TITLE
Staking gas savings + tests

### PR DESCRIPTION
**Pull request by Mvii7**

### CODE REVIEW

File: contracts/Staking.sol [hunk 1]
```diff
@@ -97,6 +97,8 @@ contract Staking is IStaking, Initializable, AccessControlUpgradeable {
     /** Variables after initial contract launch must go below here. https://github.com/OpenZeppelin/openzeppelin-sdk/issues/37 */
     /** End Variables after launch */
 
+    uint256 public basePeriod = 350;
```
Comment (Mvii7) [inline contracts/Staking.sol:100]: Does this work?

Comment (Mvii7) [reply contracts/Staking.sol:100]: Maybe we can come up with a better name?

File: test/staking/staking.ts [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+    const preUnstakeSessionData = await staking.sessionDataOf(
+      staker1.address,
+      sessionId
+    );
+
+    const interest = await staking.calculateStakingInterest(
+      preUnstakeSessionData.firstPayout, preUnstakeSessionData.lastPayout, preUnstakeSessionData.shares);
+
+    const result = await staking.getAmountOutAndPenalty(
```
Comment (Mvii7) [inline test/staking/staking.ts:171]: These 2 vars are for test only to check payout/penalty. Payout is 0 and penalty is w/e, indicating time is not advancing. Wtf is wrong?

File: contracts/Staking.sol [hunk 3]
```diff
@@ -97,6 +97,8 @@ contract Staking is IStaking, Initializable, AccessControlUpgradeable {
     /** Variables after initial contract launch must go below here. https://github.com/OpenZeppelin/openzeppelin-sdk/issues/37 */
     /** End Variables after launch */
 
+    uint256 public basePeriod;
```
Comment (Mvii7) [inline contracts/Staking.sol:100]: Lets not forget to set this

File: contracts/SubBalances.sol [hunk 4]
```diff
@@ -123,7 +128,6 @@ contract SubBalances is ISubBalances, Initializable, AccessControlUpgradeable {
                 periods[i] = _basePeriod * (i + 1);
```
Comment (Mvii7) [inline contracts/SubBalances.sol:128]: Look at this...

**Comment by Mvii7** [comment]
V1 tests will be in a separate PR, lets merge this

File: contracts/SubBalances.sol [hunk 5]
```diff
@@ -447,4 +422,8 @@ contract SubBalances is ISubBalances, Initializable, AccessControlUpgradeable {
     function setupRole(bytes32 role, address account) external onlyManager {
         _setupRole(role, account);
     }
+
+    function setStakingAddress(address stakingAddress) external onlyMigrator {
+        staking = IStaking(stakingAddress);
```
Comment (Mvii7) [inline contracts/SubBalances.sol:427]: Dont forget to set this

**Review by Dellybro — APPROVED** [review]

File: contracts/Staking.sol [hunk 6]
```diff
@@ -561,10 +572,13 @@ contract Staking is IStaking, Initializable, AccessControlUpgradeable {
...
-    
-    /** Public Setter Functions */
-    function setTotalStakedAmount(uint256 _totalStakedAmount) external onlyManager {
+
+    /** Migrator Setter Functions */
+    function setBasePeriod(uint256 _basePeriod) external onlyMigrator {
+        basePeriod = _basePeriod;
+    }
+   
+    function setTotalStakedAmount(uint256 _totalStakedAmount) external onlyMigrator {
```
Comment (Mvii7) [inline contracts/Staking.sol:581]: Changed to onlyMigrator, let me know if keep or both should be manager

Comment (Dellybro) [reply contracts/Staking.sol:581]: Should be at this point only manager

**Review by JTravaux — APPROVED** [review]

**Pull request merged**